pkg/rancher/migrations.go

```diff
@@ -39,15 +45,33 @@ func forceUpgradeLogout(configMapController controllerv1.ConfigMapController, to
 		}
 	}
 
-	// we only want to migrate if the previous installation was earlier than the migration version
+	// parse out the start and end time
+	lastMigrationStart, _ := strconv.ParseInt(cm.Data[migrationStartedKey], 10, 64)
+	lastMigrationCompleted, _ := strconv.ParseInt(cm.Data[migrationCompletedKey], 10, 64)
+
+	// we only want to migrate if the previous installation was earlier than the migration version or the last migration attempt did not complete
 	if lastMigration, ok := cm.Data[rancherVersionKey]; ok {
-		if lastMigration == "dev" || version.Compare(migrationVersion, lastMigration, "<=") {
+		if (lastMigration == "dev" || version.Compare(migrationVersion, lastMigration, "<=")) && lastMigrationStart < lastMigrationCompleted {
 			return nil
 		}
 	}
 
 	logrus.Infof("Detected %s upgrade, forcing logout for all web users", migrationVersion)
 
+	// save that we are starting a new migration
+	lastMigrationStart = time.Now().UnixNano()
```

Comment:
You get smaller numbers with just `Unix`
```suggestion
	lastMigrationStart = time.Now().Unix()
```

Reply:
I am concerned this will finish in less than a second.

Reply:
I highly doubt that, especially if you are calling out to update the config. However, just update your time check to include `=`